hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobUtils.java

```diff
@@ -907,6 +789,143 @@ public static boolean hasMobColumns(TableDescriptor htd) {
     return false;
   }
 
+  /**
+   * Get list of Mob column families (if any exists)
+   * @param htd table descriptor
+   * @return list of Mob column families
+   */
+  public static List<ColumnFamilyDescriptor> getMobColumnFamilies(TableDescriptor htd){
+
+    List<ColumnFamilyDescriptor> fams = new ArrayList<ColumnFamilyDescriptor>();
+    ColumnFamilyDescriptor[] hcds = htd.getColumnFamilies();
+    for (ColumnFamilyDescriptor hcd : hcds) {
+      if (hcd.isMobEnabled()) {
+        fams.add(hcd);
+      }
+    }
+    return fams;
+  }
+
+  /**
+   * Performs housekeeping file cleaning (called by MOB Cleaner chore)
+   * @param conf configuration
+   * @param table table name
+   * @throws IOException
+   */
+  public static void cleanupObsoleteMobFiles(Configuration conf, TableName table)
+      throws IOException {
+
+    try (final Connection conn = ConnectionFactory.createConnection(conf);
+        final Admin admin = conn.getAdmin();) {
+      TableDescriptor htd = admin.getDescriptor(table);
+      List<ColumnFamilyDescriptor> list = getMobColumnFamilies(htd);
+      if (list.size() == 0) {
+        LOG.info("Skipping non-MOB table [" + table + "]");
+        return;
+      }
+      Path rootDir = FSUtils.getRootDir(conf);
+      Path tableDir = FSUtils.getTableDir(rootDir, table);
+      // How safe is this call?
+      List<Path> regionDirs = FSUtils.getRegionDirs(FileSystem.get(conf), tableDir);
+
+      Set<String> allActiveMobFileName = new HashSet<String>();
+      FileSystem fs = FileSystem.get(conf);
+      for (Path regionPath: regionDirs) {
+        for (ColumnFamilyDescriptor hcd: list) {
+          String family = hcd.getNameAsString();
+          Path storePath = new Path(regionPath, family);
+          boolean succeed = false;
+          Set<String> regionMobs = new HashSet<String>();
+          while(!succeed) {
+            //TODO handle FNFE
+            RemoteIterator<LocatedFileStatus> rit = fs.listLocatedStatus(storePath);
+            List<Path> storeFiles = new ArrayList<Path>();
+            // Load list of store files first
+            while(rit.hasNext()) {
+              Path p = rit.next().getPath();
+              if (fs.isFile(p)) {
+                storeFiles.add(p);
+              }
+            }
+            try {
+              for(Path pp: storeFiles) {
+                HStoreFile sf = new HStoreFile(fs, pp, conf, CacheConfig.DISABLED,
+                  BloomType.NONE, true);
+                sf.initReader();
+                byte[] mobRefData = sf.getMetadataValue(HStoreFile.MOB_FILE_REFS);
+                byte[] mobCellCountData = sf.getMetadataValue(HStoreFile.MOB_CELLS_COUNT);
+                byte[] bulkloadMarkerData = sf.getMetadataValue(HStoreFile.BULKLOAD_TASK_KEY);
+                if (mobRefData == null && (mobCellCountData != null ||
+                    bulkloadMarkerData == null)) {
```

Comment:
I'm looking through `DefaultMobStoreFlusher` from before this patch to confirm that we'll recognize old files properly and I think this works because all three of these hfile metadata entries will be null. Is that correct?

Reply:
It turned out that MOB_CELLS_COUNT is irrelevant here because it is the meta attribute of MOB file only and here we go through regular store files. I changed the logic around this and how we store MOB_FILE_REFS during flush and compactions. The old code ignored this attribute if number of references was 0. The new code always stores this meta key, even if number of references is 0, thus giving us  a way to distinguish that the store file was created by new code.